### ISSUE TITLE
Make drush_scan_directory() ignored directories configurable

### ISSUE THREAD

**Issue by ademarco** [drush-ops/drush#3199]
Work done in #2062 excluded `node_modules` and `bower_components` in `drush_scan_directory()`:

```php
$nomask = array_merge($nomask, array('node_modules', 'bower_components'));
```

We could make this configurable with a global Drush option that defaults to the values above, like:

```php
$options['ignored-directories'] = ['node_modules', 'bower_components'];
```

Would this be something interesting? I could work on it if so.

**Comment by drupol** [comment]
:+1:

**Comment by greg-1-anderson** [comment]
What are the use cases for this?  In Drush 9, we simply removed deep scanning for command files and aliases. It seems that limiting scan depth (via option) would be a better place to start than customizing the exclusions. Even better would be to prevent overlap between the places Drush is scanning, and the places where you have large directories.

What are the large directories that you want to exclude, and where are they located in your Drupal site's codebase?

**Comment by ademarco** [comment]
So the use-case is pretty much what described here: https://github.com/nuvoleweb/drupal-component-scaffold

We want to develop Drupal modules and themes as any other modern PHP library is developed, meaning:

1. Clone module repository
2. Rune composer install: this downloads all development dependencies which, in this case, is Drupal core, required modules etc. and build a fully functional test Drupal site within the module's directory
3. Finally: symlink repository root, (i.e. the module/theme directory) in the right place within the built target site, and here is where recursion happens.

The patch above would allow us to exclude the build directory avoiding Drush to recurse following the symlink to the very same repository root.

The approach described above has several benefits, like:

- Lean travis (or any other) CI configuration files
- Allow all module's maintainers / contributors to easily bootstrap the same development environment used on CI lowering the barrier to work with tests etc.

For a working example of such an approach check the [UI Patterns module](https://www.drupal.org/project/ui_patterns), especially its [travis configuration file](https://github.com/nuvoleweb/ui_patterns/blob/8.x-1.x/.travis.yml) and [composer.json](https://github.com/nuvoleweb/ui_patterns/blob/8.x-1.x/composer.json).

**Comment by weitzman** [comment]
Wow, drupal-component-scaffold looks familiar. Thats the same approach that Drush itself uses in its unit tests. We build a "Site Under Test" and symlink Drush into it via Composer's [Path repository](https://getcomposer.org/doc/05-repositories.md#path). One minor difference is that the Site Under Test doesn't sit within Drush, its an outside directory. That avoids the recursion problem. Anyway, I 100% support the goals of drupal-component-scaffold.

As Greg says, this may not be needed for Drush9. Would be great if you could put Drush9 into drupal-component-scaffold. It hit RC1 today and is already the recommended version for Drupal 8 sites. If you still are seeing the recursion, then fixing it via an option/config sounds OK.

**Comment by greg-1-anderson** [comment]
https://github.com/nuvoleweb/drupal-component-scaffold looks super useful.

**Comment by greg-1-anderson** [comment]
Maybe if drupal-component-scaffold could be refactored to keep the SUT in a separate location, then we could utilize it in the Drush tests. That would keep our tests clean and maintainable, and would also be a really good guarentee that drupal-component-scaffold is always working with Drush.

**Comment by ademarco** [comment]
Happy to hear you like `drupal-component-scaffold`! It is indeed the exact same approach Drush is using to test itself and I'd love to see `drupal-component-scaffold` being used in Drush testing.

Shall I open an issue about that (i.e. drupal-component-scaffold and Drush 9+ testing) so we can investigate further?

About Drush 9 and recursion: I can indeed confirm that we don't have the issue on 9.x: I've used Drush 9 on UI Patterns together with `drupal-component-scaffold` and I was able to install the site without any problem.

Still it's probably cleaner to provide an option to exclude specific directories (with sensitive defaults like `node_modules` and `bower_components`) instead of hardcoding them within the code.

But since the issue does not manifest itself then we can surely drop it for version 9!

The situation is different on Drush 8.x, though:

- In our organisation we still develop and maintain modules and themes in Drupal 7 where, unfortunately, we can't use Drush 9 and we would like to use a similar approach as in `drupal-component-scaffold`
- Drush 9 supports Drupal 8.4+ and, in some cases, it might be desirable to test a module on Drupal 8.3 too (UI Patterns does that): in this case I'm still forced to use Drush 8.x

All in all I still see a use case for this patch on 8.x, shall I maybe open a PR on that branch?
